Add a rake task to cleanup stale database records

NobodysNightmare opened the pull request:
### Summary

Based on discussions in https://github.com/doorkeeper-gem/doorkeeper/issues/536,
this merge request wraps "safe to execute" maintenance tasks
in a rake task.

For now this is basically the one-liner provided by @tute, but I assume that there is more
data that we can safely remove during such a cleanup.

My main motivation is that migrations on the `oauth_access_tokens` table tend to take quite a long
time, when that table has grown rather large.

NobodysNightmare commented on lib/doorkeeper/rake/setup.rake:
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+namespace :doorkeeper do
+  task setup: :environment do
```
This is inspired by `resque`, where you have a hook that is executed before each rake task of that gem.

I decided to depend on `:environment` by default, but to be honest I never used Doorkeeper outside a Rails-context. If it is neccessary, we could also make the dependency on `environment` optional, as they've done in `resque`.

nbulaj replied:
Looks pretty good for me :+1:

NobodysNightmare commented on README.md:
````diff
@@ -243,6 +244,27 @@ You may want to check other ways of authentication
+the following line to your `Rakefile`:
+
+```ruby
+Doorkeeper::Rake.load_tasks
````
Alternatively we could load all the rake tasks when requiring `doorkeeper`, but I feel that this would force people to use rake, when it is not a strict dependency of this gem.

nbulaj replied:
Yep, everything is OK

NobodysNightmare commented on README.md:
````diff
@@ -243,6 +244,27 @@ You may want to check other ways of authentication
+rake doorkeeper:db:cleanup
+```
+
+Note that by default those records are retained to allow stricter auditing.
````
Actually I assume that this is the reason. Feel free to fill me in on a better explanation :)

NobodysNightmare commented:
@tute wrote:
> Doorkeeper keeps revoked tokens, you can delete them with a query like:
> ```
> DELETE FROM oauth_access_tokens WHERE revoked_at IS NOT NULL AND revoked_at < NOW();
> ```

On our database I can see that this does not hit the majority of tokens. We have lots of tokens, that simply expired (`created_at = 2016-...`, `expires_in = 7200`).

Is it safe to delete those after a short hold period? Should I keep them? If yes, why?

nbulaj commented:
Hi @NobodysNightmare . What about to remove all the tokens / grants not for current time, but without token TTL?

```ruby
.where(tokens[:revoked_at].lt(Time.current - Doorkeeper.configuration.access_token_expires_in))
```

It will prevent from removing still-valid tokens. What do you think?

NobodysNightmare commented:
Hey @nbulaj, I assume you mean tokens that were **created** more than the TTL amount ago? Your code example refers to the revocation date. I.e.:

```ruby
.where(tokens[:created_at].lt(Time.current - Doorkeeper.configuration.access_token_expires_in))
```

Your approach seems nice to me and should be easy to implement. I am a little afraid of the edge case, where the TTL of tokens is being changed. For that purpose the table contains a column for the ttl of the token. So I could build a SQL query that calculates whether `created_at + ttl < Time.current`. However, calculations with dates are specific to your SQL backend (MySQL is different from Postgres and so on).

Do you think it is sufficient to rely on `Doorkeeper.configuration.access_token_expires_in`? I can implement that.

nbulaj commented:
Yep, sorry, I just copied your example :)

It's ok to remove revoked tokens, but also we can remove not revoked, but already not accessible tokens (Time.current - TTL).

> However, calculations with dates are specific to your SQL backend (MySQL is different from Postgres and so on).

But we use ORM's that must implement it for us. isn' it? :)

NobodysNightmare commented:
I just added a commit that also cleans up expired tokens and grants using the simple approach that you proposed.

I split the rake tasks further, so that anyone would be able to just run the cleanups they need.

> The only edge case is custom expiration time :\ I can't say right now if it can be used for rake tasks. What do you think?

I chose the static approach now (reading TTL from configuration) and also added a disclaimer for that to the README.

Alternative approaches:
* Let the SQL backend decide which records are matching and delete them
    * this would be fast (just one `DELETE FROM` query)
    * I have no clue how to do this portable, since the `where` clause had to contain some DBMS specific date calculation (e.g. for MySQL: `where('DATE_ADD(created_at, INTERVAL expires_in SECOND) < ?', Time.current)`)
        * this is where my ORM might help me, but I don't know of the relevant ActiveRecord feature
* implement it using expiration calculation in Ruby (`Doorkeeper::AccessToken.select(&:expired?).each(&:destroy)`)
    * slow N+1 query
    * more accurate than current implementation

nbulaj commented on lib/doorkeeper/rake/db.rake:
```diff
@@ -0,0 +1,51 @@
+  namespace :db do
+    desc 'Removes stale data from doorkeeper related database tables'
+    task cleanup: [
+                    'doorkeeper:db:cleanup:revoked_tokens',
```
Let's indent it (all the array items) with just 2 spaces and final `]` under the `t` of `task`

nbulaj commented on lib/doorkeeper/rake/db.rake:
```diff
@@ -0,0 +1,51 @@
+    desc 'Removes stale data from doorkeeper related database tables'
+    task cleanup: [
+                    'doorkeeper:db:cleanup:revoked_tokens',
+                    'doorkeeper:db:cleanup:expired_tokens',
```
Let's indent it (all the array items) with just 2 spaces and final `]` under the `t` of `task`

nbulaj commented:
Also could you please squash all your commits to a single one after you finish? Thanks!

NobodysNightmare commented on lib/doorkeeper/orm/active_record/stale_record_cleaner.rb:
```diff
@@ -0,0 +1,26 @@
+        def clean_expired(ttl)
+          table = @model.arel_table
+          @model.where(table[:created_at].lt(Time.current - ttl))
+                .delete_all
```
Is @houndci-bot right on this one?

The rubocop default is having leading dots and I don't see doorkeeper overriding that...

nbulaj replied:
Just ignore it. Not a big deal.

nbulaj replied:
Just ignore it. Not a big deal.

NobodysNightmare commented on lib/doorkeeper/rake/db.rake:
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+namespace :doorkeeper do
```
I think that rake namespaces should not be affected by this, while I agree that `task` should not be too long.

We are using a rubocop config like this at our projects:

```yml
# We need to configure exemptions for blocks that we generally accept to be
# long, since they are less comparable to methods and more comparable to
# modules/classes.
Metrics/BlockLength:
  ExcludedMethods:
    - context
    - describe
    - namespace
    - factory
    - define
    - shared_examples_for
    - shared_examples
    - xcontext
    - xdescribe
```

What do you think?

nbulaj replied:
I just disable this check for all RSpec suite because it is too noisy :) Just ignore this one

NobodysNightmare commented:
Addressed your points and squashed to a single commit.

I extracted the beefy part of the rake task into a service class. This also helped in reducing the duplication that was present between the tasks.

I am a little confused by the hound results, see my responses above :)

NobodysNightmare commented on spec/lib/orm/active_record/stale_record_cleaner_spec.rb:
```diff
@@ -0,0 +1,77 @@
+require 'spec_helper'
```
fixed

nbulaj commented on lib/doorkeeper/orm/active_record/stale_record_cleaner.rb:
```diff
@@ -0,0 +1,26 @@
+module Doorkeeper
+  module Orm
+    module ActiveRecord
+      class StaleRecordCleaner
```
Just a small hint: maybe better to name it '`StaleRecordsCleaner`' (as we remove more than one record)? :)

NobodysNightmare replied:
True :)

fixed

NobodysNightmare commented:
@nbulaj I was thinking about this MR again and I am not sure anymore, that deleting expired access tokens is a good idea.

As far as I can tell, these tokens are still used (after their expiry) to obtain new access tokens using their `refresh_token`.

By cleaning expired tokens, we basically also invalidate the corresponding `refresh_tokens`... After which amount of time should this be considered a safe operation? As far as I can tell, the refresh tokens are so far never expiring...

nbulaj commented:
Hey, @NobodysNightmare . We completely forgot about refresh tokens, you are right! But it is a configurable option (`use_refresh_token`), so by default there are no refresh tokens on OAuth2 provider with Doorkeeper. Therefore this task can be helpful in this case.

What I can suggest is to add configuration check to rake tasks that will see if this option is activated. If yes, it must: **1)** stop working with the warning message, **2)** show warning message and ask if user is sure he want's to remove the tokens and if yes - continue the operation.

What do you think?

NobodysNightmare commented:
Warning and having an option to continue anyways sounds good... What should we do about the general clean task then? (the one that executes all four cleanups)

Failing it completely just when one independent step does not work sounds not so nice...

I see three options:
* remove the common task and just expose the separate tasks
* skip expired tokens in the common task when it is not safe
    * log that it was skipped
* ~~fail common task, but ensure expired tokens are handled last~~

edit: Thinking about it, the last option is probably lazy and not useful for anyone...

NobodysNightmare commented:
Of course one additional idea:

Even if `use_refresh_token` is true, not all access tokens will have a refresh token associated. When obtaining a token via client credentials flow, there is never a refresh token associated.

So alternatively, we could only delete expired refresh tokens that also have **no** refresh token.

For our database this would still help a lot, since we have many more tokens issued via client credentials flow...

It would also be a solution that has the rake task always behave the same. We are always removing all the expired tokens that have surely no purpose anymore...

nbulaj commented:
To skip expired tokens and to log that is a good idea I think. And every developer can then run another task that will destroy expired tokens.

However, this one sounds better to me:

> we could only delete expired refresh tokens that also have **no** refresh token.

Other cases depends on the purposes of the project, so I think every developer can write his own tasks.

NobodysNightmare commented:
> we could only delete expired refresh tokens that also have **no** refresh token.

Implemented that proposal here: https://github.com/doorkeeper-gem/doorkeeper/pull/1080

lefty313 commented on lib/doorkeeper/orm/active_record/stale_records_cleaner.rb:
```diff
@@ -0,0 +1,26 @@
+                .delete_all
+        end
+
+        def clean_expired(ttl)
```
@nbulaj first of all sorry for commenting already merged pull request. 

IMO people will use this rake task when database is already bloated but in such case it can easily kill production database because doorkeeper does not generate index for `AccessToken#created_at` field. Maybe we should consider adding this index by default or add warning in `Readme.md` ?

nbulaj replied:
@NobodysNightmare it definitely makes sense. Maybe `find_each` could help us. Would you like to produce a PR?

nbulaj replied:
Fixed in https://github.com/doorkeeper-gem/doorkeeper/pull/1234

NobodysNightmare replied:
Sorry for not responding. I had this somewhere in my inbox with "I want to answer there" :)

Cool that you already went ahead and fixed it :+1:
